# Homework 1

## FINM 36700 - 2024

### UChicago Financial Mathematics

* Mark Hendricks
* hendricks@uchicago.edu

## HBS Case

### *The Harvard Management Company and Inflation-Indexed Bonds*

### Notation
(Hidden LaTeX commands)

$$\newcommand{\mux}{\tilde{\boldsymbol{\mu}}}$$
$$\newcommand{\wtan}{\boldsymbol{\text{w}}^{\text{tan}}}$$
$$\newcommand{\wtarg}{\boldsymbol{\text{w}}^{\text{port}}}$$
$$\newcommand{\mutarg}{\tilde{\boldsymbol{\mu}}^{\text{port}}}$$
$$\newcommand{\wEW}{\boldsymbol{\text{w}}^{\text{EW}}}$$
$$\newcommand{\wRP}{\boldsymbol{\text{w}}^{\text{RP}}}$$
$$\newcommand{\wREG}{\boldsymbol{\text{w}}^{\text{REG}}}$$

***

# 1. HMC's Approach

**Section 1 is not graded**, and you do not need to submit your answers. But you are encouraged to think about them, and we will discuss them.

### 1. 
There are thousands of individual risky assets in which HMC can invest.  Explain why MV optimization across 1,000 securities is infeasible.


The MV solution requires us to invert the covariance matrix: 
. The term is highly unstable for large number of securities, giving also a unstable result. With 1,000 securities, the estimation error of means, volatilities, and covariances produce impactful errors in allocation. The latter happens due to the high sensitivity to changes in input data. If the input data in the out-of-sample is slightly different from the in-sample, the out-of-sample optimal portfolio will have considerably different weights from the in-sample.

Optimizing for a smaller number of assets reduces the allocation consequence of estimation errors in the moments of the assets.

1. Feasibility of MV Optimization Across 1,000 Securities
MV optimization across 1,000 securities is impractical due to computational challenges and the minuscule weight each security would need to hold in the portfolio. Optimizing such a large number involves an extensive number of variables (weights) and an enormous set of covariances to consider, which increases computational complexity exponentially. Additionally, the practical aspect of acquiring such fractional shares of securities where weights approach near zero becomes unfeasible in real market conditions, due to minimum trade sizes and transaction costs.

2. Conditions for Effective Two-Stage Optimization
In two-stage optimization, the first stage involves optimizing within individual asset classes, and the second stage involves optimizing across these classes. For this approach to effectively approximate a full MV optimization, the securities within each asset class must exhibit higher intraclass correlations compared to inter-class correlations. This low inter-class correlation ensures that each asset class behaves almost independently, making local optimizations within asset classes a good proxy for part of a global solution. If the simplex algorithm or another linear programming method is used in optimization, it tends to select corner solutions where some weights can be zero, making the effective partitioning critical for achieving near-optimal global results.

3. Classification of TIPS as a Separate Asset Class
Considering TIPS' unique characteristics, especially their inflation protection capability and distinct risk-return profile indicated by a Sharpe ratio greater than 1, they should form their own asset class. Their low correlation with other asset classes further justifies this separation, allowing for more targeted inflation risk management within portfolio optimizations. This distinct classification aids in precise control over exposure to inflation-related risks, which is not typically achievable when grouped with other fixed-income securities.

4. Focus on Real Returns in Portfolio Allocation
Focusing on real returns, which account for inflation, is critical because nominal returns can misrepresent the true economic value of portfolio gains by not considering the erosion of purchasing power. Real returns provide a more accurate basis for comparing the performance of different asset classes under varying inflationary conditions. The use of real returns affects the mean-variance optimization by altering the expected returns and covariance matrix used in the calculations, potentially leading to different optimal solutions than those based on nominal returns. The impact of inflation can particularly skew the risk assessments and correlations between assets, thus real returns offer a more stable foundation for MV optimization.

5. Adjusting the Optimization Problem to Reflect Bounded Solutions
To model the constraints used by Harvard, as depicted in Exhibits 5 and 6, the optimization problem should incorporate bounded inequality constraints for each asset class, limiting allocations within specified ranges (e.g., 0-100%). Additional constraints might specify narrower ranges or special conditions like those for cash (e.g., -50% to 100%). These modifications ensure the optimizer adheres to policy limits and practical investment guidelines, avoiding extreme allocations that could lead to undesirable risk exposures or liquidity issues. Integrating such constraints might involve using the simplex algorithm which optimizes at vertices of the feasible region, typically resulting in allocations being pushed to their bounds.

6. Zero Allocation to Domestic Equities and Bonds
The zero allocation to domestic equities and bonds might indicate that these asset classes, under the constraints imposed, do not adequately contribute to the desired portfolio outcomes across the target return spectrum. This outcome can be attributed to the simplex algorithm, which optimizes at the extreme points of the feasible region, effectively pushing some asset weights to zero when they do not sufficiently enhance the portfolio's risk-return profile. Conceptually, in an unconstrained optimization, there might be small allocations to these classes if they offered some diversification or return benefit without significantly increasing risk.

7. Impact of Tightened Constraints on the Mean-Variance Tradeoff
The tightening of allocation constraints as seen in Exhibit 6 leads to a noticeable deterioration in the mean-variance tradeoff, where the Sharpe ratio declines from 0.38 to about 0.35 or 0.36, reflecting a roughly 5% reduction in portfolio efficiency. This change indicates that tighter constraints restrict the optimizer's ability to allocate capital to optimally balance risk and return, thus reducing the overall effectiveness of the portfolio in achieving higher risk-adjusted returns. This effect is particularly pronounced when the optimizer is constrained from allocating to assets that might otherwise improve the portfolio's performance under different risk-return objectives.


### 2.
Rather than optimize across all securities directly, HMC runs a two-stage optimization.
1. They build asset class portfolios with each one optimized over the securities of the specific asset class.  
2. HMC combines the asset-class portfolios into one total optimized portfolio.

In order for the two-stage optimization to be a good approximation of the full MV-optimization on all assets, what must be true of the partition of securities into asset classes?


For the two-stage optimization to be a good approximation securities within each of the asset classes must be highly correlated while securities of different assets classes must have lower correlation. In other words, within a group of securities, return characteristics must be highly homogeneous and asset classes must be sufficiently distinct.

If those two conditions are met, the two-stage process can approximate well the full MV-optimization.


### 3.
Should TIPS form a new asset class or be grouped into one of the other 11 classes?


There are arguments for both sides.
Favorable arguments for TIPS being a separate asset class:

TIPS provide a high diversification component, having lower correlation to bonds than SPY has to other equity buckets.
TIPS returns are a direct protection to inflation, having a priori a different return characteristic when compared to any other asset class.
Favorable arguments for TIPS NOT being a separate asset class:

TIPS have a priori a similar default risk to public bonds.
Adding TIPS mostly impacts the allocation in domestic bonds, which favours the idea that it could be just another asset in this bucket.


### 4. 
Why does HMC focus on real returns when analyzing its portfolio allocation? Is this just a matter of scaling, or does using real returns versus nominal returns potentially change the MV solution?


Using real returns in the MV optimization process, therefore, does not merely change the scale of returns but alters the fundamental inputs into the decision-making process by adjusting for the loss of purchasing power due to inflation. This adjustment ensures that the optimization process targets returns that genuinely increase the investor's wealth in real terms, aligning investment decisions more closely with actual financial goals and risk preferences. HMC prioritizes real returns over nominal returns in portfolio analysis because real returns account for inflation, providing a more accurate measure of purchasing power and true investment performance.
Furthermore, the closed form solution of the MV frontier to find the tangency portfolio only works when using excess-returns:If this formula is used with returns as inputs, it produces the point of maximum mean divided by variance, not equal to the point of maximum Sharpe ratio.


### 5.
The case discusses the fact that Harvard places bounds on the portfolio allocation rather than implementing whatever numbers come out of the MV optimization problem.

How might we adjust the stated optimization problem in the lecture notes to reflect the extra constraints Harvard is using in their bounded solutions given in Exhibits 5 and 6?


With the new constrains, the problem has no closed formula solution.

In practice, MV frontier is widely used with constraints. While the formula leads to less extreme weights, it is highly sensitive to the constraint boundaries and requires the investor to set a great amount of arbitrary parameters.

More recently, Machine Learning algorithms of regularization have gained traction in MV optimization: RIDGE and LASSO and Elastic Net also produce less extreme weight allocation and do not require investors to set arbitrary boundaries.


### 6. 
Exhibits 5 shows zero allocation to domestic equities and domestic bonds across the entire computed range of targeted returns, (5.75% to 7.25%). Conceptually, why is the constraint binding in all these cases? What would the unconstrained portfolio want to do with those allocations and why?


In Exhibit 5, assuming certain conditions, an MV optimization strategy involves shorting domestic equities and domestic bonds to increase exposure to other long positions, particularly TIPS, due to their attractive Sharpe Ratio and hedging properties. This strategy is consistent across various expected return targets.


### 7.
Exhibit 6 changes the constraints, (tightening them in most cases.) How much deterioration do we see in the mean-variance tradeoff that Harvard achieved?

In Exhibit 6, with stricter constraints, achieving any expected return target requires higher volatility. Consequently, the Mean/Volatility Ratios decrease from 0.38 to 0.35 in most cases. This signifies a nearly 10% decline in optimization performance when transitioning from Exhibit 5 to Exhibit 6 due to increased constraints. An unconstrained solution would yield a Mean/Volatility Ratio exceeding 0.80. For expected return targets further from the tangency point, this decline in performance worsens as the MV-efficient portfolios cannot be attained through appropriate short positions.

***

# 2 Mean-Variance Optimization

<i>This section is graded for a good-faith effort by your group. Submit your write-up- along with your supporting code. </i>

### Data
You will need the file in the github repo, `data/multi_asset_etf_data.xlsx`.
- The time-series data gives monthly returns for the 11 asset classes and a short-term Treasury-bill fund return, ("SHV",) which we consider as the risk-free rate.
- The data is provided in total returns, (in which case you should ignore the SHV column,) as well as excess returns, (where SHV has been subtracted from the other columns.)
- These are nominal returns-they are not adjusted for inflation, and in our calculations we are not making any adjustment for inflation.
- The exhibit data that comes via Harvard with the case is unnecessary for our analysis.

### Model
We are going to analyze the problem in terms of **excess** returns.
- Thus, you will focus on the `Excess Returns` section of the lecture notes, especially the formulas on slide 50.
- Be sure to use the`excess returns` tab of the data.

### Format
In the questions below, **annualize the statistics** you report.
- Annualize the mean of monthly returns with a scaling of 12.
- Annualize the volatility of monthly returns with a scaling of $\sqrt{12}$
- The Sharpe Ratio is the mean return divided by the volatility of returns. Accordingly, we can annualize the Sharpe Ratio with a scaling of $\sqrt{12}$
- Note that we are not scaling the raw timeseries data, just the statistics computed from it (mean, vol, Sharpe). 

### Footnotes

#### Data File
* The case does not give time-series data, so this data has been compiled outside of the case, and it intends to represent the main asset classes under consideration via various ETFs. For details on the specific securities/indexes, check the “Info” tab of the data.

#### Risk-free rate
* In the lecture-note we considered a constant risk-free rate. It is okay that our risk-free rate changes over time, but the assumption is that investors know it’s value one-period ahead of time. Thus, at any given point in time, it is a risk-free rate for the next period. (This is often discussed as the "bank account" or "money market account" in other settings.

## 1. Summary Statistics
* Calculate and display the mean and volatility of each asset’s excess return. (Recall we use volatility to refer to standard deviation.)
* Which assets have the best and worst Sharpe ratios? Recall that the Sharpe Ratio is simply the ratio of the mean-to-volatility of excess returns:
$$\text{sharpe ratio of investment }i = \frac{\mux_i}{\sigma_i}$$

## 2. Descriptive Analysis
* Calculate the correlation matrix of the returns. Which pair has the highest correlation? And the lowest?
* How well have TIPS done in our sample? Have they outperformed domestic bonds? Foreign bonds?

## 3. The MV frontier.
* Compute and display the weights of the tangency portfolios: $\wtan$.
* Does the ranking of weights align with the ranking of Sharpe ratios?
* Compute the mean, volatility, and Sharpe ratio for the tangency portfolio corresponding to
$\wtan$.

## 4. TIPS
Assess how much the tangency portfolio (and performance) change if...
* TIPS are dropped completely from the investment set.
* The expected excess return to TIPS is adjusted to be 0.0012 higher than what the historic sample shows.

Based on the analysis, do TIPS seem to expand the investment opportunity set, implying that Harvard should consider them as a separate asset?

***

In [1]:
import pandas as pd
import polars as pl
import numpy as np
import os 

In [2]:
os.getcwd()

'c:\\Users\\Andik\\OneDrive\\Desktop\\Chicago\\Portfolio theory and risk management\\githubrepo\\finm-portfolio-2024\\homework'

In [3]:
# Load data
data_path = "../data/multi_asset_etf_data.xlsx"

In [4]:
data = pd.read_excel(data_path,sheet_name="excess returns")
if data.index.name != "Date":
    data.set_index("Date",inplace=True)
data.head()


BWX       DBC       EEM       EFA       HYG       IEF  \
Date                                                                     
2011-02-28  0.006527  0.041620 -0.000527  0.035408  0.014264 -0.002174   
2011-03-31  0.008255  0.026409  0.062723 -0.024055  0.000251 -0.001719   
2011-04-30  0.048261  0.045015  0.026783  0.055715  0.015432  0.017868   
2011-05-31 -0.013447 -0.051625 -0.029301 -0.021962  0.001432  0.025156   
2011-06-30 -0.000270 -0.042817 -0.009528 -0.012280 -0.005878 -0.005217   

                 IYR       PSP       QAI       SPY       TIP  
Date                                                          
2011-02-28  0.045113  0.040056  0.002493  0.034647  0.007118  
2011-03-31 -0.011107  0.016495  0.005348 -0.000052  0.011730  
2011-04-30  0.046089  0.058127  0.018490  0.028418  0.023236  
2011-05-31  0.010233 -0.041465  0.000099 -0.011115  0.002758  
2011-06-30 -0.031233 -0.042754 -0.010948 -0.017042  0.007377

In [5]:
## Summary Statistics
from scipy.stats import describe

def summary_stats(data):

    data_mean = describe(data,axis = 0)[2] * 12
    data_sample_var = describe(data,axis = 0)[3] 
    data_sample_stdev = describe(data,axis = 0)[3] **0.5 * 12**0.5
    data_sharpe =  data_mean / data_sample_stdev 
    # print(f"Securities names: {data.columns}")
    # print(f"Mean excess returns of each securities: {data_mean}")
    # print(f"Var of securities: {data_sample_var}")
    # print(f"Standard deviations of securities: {data_sample_stdev}")
    # print(f"Sharpe ratio for each security: {data_sharpe}")

    result_1 = pd.DataFrame({"Asset": data.columns,
                            "Mean returns": data_mean,
                            "Stddev":data_sample_stdev,
                            "Sharpe": data_sharpe})

    # data validation
    mean_check = data.apply(lambda x: x.mean(), axis=0).to_numpy()

    stdev_check = data.apply(lambda x: np.std(x, ddof=1), axis=0).to_numpy()

    val1 = np.isclose(mean_check*12,data_mean,rtol=0.01).all()
    val2 = np.isclose(stdev_check*12**0.5,data_sample_stdev,rtol=0.01).all()
    if val1 and val2: 
        print(f"Security with best sharpe ratio: {data.columns[np.argmax(data_sharpe)]}")
    else:
        print("Error in calculations")
    return result_1

result_1 = summary_stats(data)
display(result_1)

Security with best sharpe ratio: SPY


Asset  Mean returns    Stddev    Sharpe
0    BWX     -0.011888  0.081671 -0.145563
1    DBC     -0.009086  0.168455 -0.053935
2    EEM      0.026960  0.179940  0.149829
3    EFA      0.055037  0.152203  0.361605
4    HYG      0.037356  0.077289  0.483335
5    IEF      0.013939  0.063197  0.220561
6    IYR      0.077912  0.169585  0.459426
7    PSP      0.092851  0.215238  0.431386
8    QAI      0.014959  0.049007  0.305241
9    SPY      0.126983  0.143066  0.887578
10   TIP      0.016844  0.051258  0.328618

In [6]:
## Descriptive Analysis

### Correlation matrices
if data.index.name != "Date":
    data.set_index("Date",inplace=True)
correlation_matrix = data.corr(method="pearson")

print("Correlation Matrix:")
display(correlation_matrix)




Correlation Matrix:


BWX       DBC       EEM       EFA       HYG       IEF       IYR  \
BWX  1.000000  0.233222  0.638705  0.600620  0.620946  0.568903  0.562599   
DBC  0.233222  1.000000  0.516908  0.521447  0.471924 -0.300523  0.290040   
EEM  0.638705  0.516908  1.000000  0.820770  0.694174  0.020477  0.589003   
EFA  0.600620  0.521447  0.820770  1.000000  0.790742  0.018537  0.705926   
HYG  0.620946  0.471924  0.694174  0.790742  1.000000  0.180087  0.739567   
IEF  0.568903 -0.300523  0.020477  0.018537  0.180087  1.000000  0.297047   
IYR  0.562599  0.290040  0.589003  0.705926  0.739567  0.297047  1.000000   
PSP  0.552010  0.467111  0.761796  0.910596  0.814182  0.017486  0.753810   
QAI  0.662425  0.494467  0.791130  0.863957  0.811161  0.176287  0.718397   
SPY  0.466152  0.448389  0.699939  0.864822  0.794782 -0.005053  0.760646   
TIP  0.680001  0.114379  0.379465  0.386677  0.541758  0.745025  0.590805   

          PSP       QAI       SPY       TIP  
BWX  0.552010  0.662425  0.466152  0.680001  
DBC  0.467111  0.494467  0.448389  0.114379  
EEM  0.761796  0.791130  0.699939  0.379465  
EFA  0.910596  0.863957  0.864822  0.386677  
HYG  0.814182  0.811161  0.794782  0.541758  
IEF  0.017486  0.176287 -0.005053  0.745025  
IYR  0.753810  0.718397  0.760646  0.590805  
PSP  1.000000  0.870673  0.890719  0.415854  
QAI  0.870673  1.000000  0.862479  0.527724  
SPY  0.890719  0.862479  1.000000  0.393647  
TIP  0.415854  0.527724  0.393647  1.000000

In [7]:
def find_corr(correlation_matrix,type: str = "min"):    
    

    if type == "min":
        matrix = np.absolute(correlation_matrix.to_numpy())
        val = matrix.min()
        corr = np.where(matrix == val)
    elif type=="max":
        matrix = correlation_matrix.to_numpy() - np.eye(correlation_matrix.shape[0])
        val =matrix.max()
        corr = np.where(matrix == val)
    else:
        raise ValueError
    row = corr[0][0]
    col = corr[0][1]
    row_name = correlation_matrix.index[row]
    col_name =  correlation_matrix.index[col]
    print(f"{type.capitalize()} absolute correlation is between '{row_name}' and '{col_name}' with a coefficient of {val:.7f}")
    return row_name, col_name, val

find_corr(correlation_matrix, "min")
find_corr(correlation_matrix, "max")





Min absolute correlation is between 'IEF' and 'SPY' with a coefficient of 0.0050534
Max absolute correlation is between 'EFA' and 'PSP' with a coefficient of 0.9105957


('EFA', 'PSP', 0.9105956995777845)

Based on the updated dataset, the performance of TIPS (Treasury Inflation-Protected Securities) relative to domestic bonds (IEF) and foreign bonds (BWX) shows a significant advantage for TIPS. With a mean return of 0.016844 and a Sharpe ratio of 1.138368, TIPS notably outperform both bond categories. In comparison, IEF, representing domestic bonds, offers a mean return of 0.013939 and a Sharpe ratio of 0.764046, while BWX, standing for foreign bonds, shows a negative mean return of -0.011888 and a negative Sharpe ratio of -0.504246. This analysis highlights TIPS as not only providing higher returns but also maintaining a more favorable risk-adjusted return ratio than both domestic and foreign bond investments within this sample. The stark contrast in performance, especially in terms of Sharpe ratios, underscores the effectiveness and robustness of TIPS as an investment option under current market conditions.

In [ ]:
## The MV frontier
from numpy import linalg

def tangency_portfolio(data: pd.DataFrame):
    """
    Calculate the weights of the tangency portfolio using mean excess returns and a returns DataFrame.
    `data` should be a DataFrame where rows are observations and columns are different assets.
    scipy.stats.describe must be present
    """
    data_mean = describe(data,axis = 0)[2] * 12
    securities = data.columns
    mean_excess_returns = np.reshape(data_mean, (-1,))
    covariance_matrix = data.cov(ddof=1).to_numpy()
    inverse_cov_matrix = linalg.inv(covariance_matrix)
    tangency_portfolio = (inverse_cov_matrix @ mean_excess_returns) / (np.ones(shape=(1, inverse_cov_matrix.shape[0])) @ inverse_cov_matrix @ mean_excess_returns)
    
    if not (round(tangency_portfolio.sum()) == 1):
        raise ValueError("The sum of the tangency portfolio weights does not equal 1.")
    weights_df = pd.DataFrame({"Asset": securities,
                               "Weights":tangency_portfolio})
    return weights_df

display(tangency_portfolio(data=data))


Asset   Weights
0    BWX  2.730253
1    DBC -0.111241
2    EEM -0.615698
3    EFA  0.530692
4    HYG -0.832578
5    IEF -4.682480
6    IYR  0.877108
7    PSP  0.277002
8    QAI  7.220592
9    SPY -4.904086
10   TIP  0.510436

In [9]:
print('Sorted Rankings for weights')
display(tangency_portfolio(data=data).sort_values(by = "Weights", ascending=False,axis =0))
print("Rankings for Sharpe")
display(result_1.sort_values(by = "Sharpe",ascending=False,axis=0))



if (result_1.sort_values(by = "Sharpe",ascending=False,axis=0).index == tangency_portfolio(data=data).sort_values(by = "Weights", ascending=False,axis =0).index).all():
    print(f"Rankings are same")
else:
    print(f"Rankings are different")


Sorted Rankings for weights


Asset   Weights
8    QAI  7.220592
0    BWX  2.730253
6    IYR  0.877108
3    EFA  0.530692
10   TIP  0.510436
7    PSP  0.277002
1    DBC -0.111241
2    EEM -0.615698
4    HYG -0.832578
5    IEF -4.682480
9    SPY -4.904086

Rankings for Sharpe


Asset  Mean returns    Stddev    Sharpe
9    SPY      0.126983  0.143066  0.887578
4    HYG      0.037356  0.077289  0.483335
6    IYR      0.077912  0.169585  0.459426
7    PSP      0.092851  0.215238  0.431386
3    EFA      0.055037  0.152203  0.361605
10   TIP      0.016844  0.051258  0.328618
8    QAI      0.014959  0.049007  0.305241
5    IEF      0.013939  0.063197  0.220561
2    EEM      0.026960  0.179940  0.149829
1    DBC     -0.009086  0.168455 -0.053935
0    BWX     -0.011888  0.081671 -0.145563

Rankings are different


In [10]:

portfolio_mean_excess_returns_df = tangency_portfolio(data=data)
merged_df =pd.merge(left=result_1,right=portfolio_mean_excess_returns_df,on="Asset", how="outer")
tangency_portfolio_annualised_mean = merged_df.eval("`Mean returns` * Weights").sum()
covariance_matrix = data.cov(ddof=1).to_numpy()
tangency_portfolio_variance = merged_df["Weights"].to_numpy().T @ covariance_matrix @  merged_df["Weights"].to_numpy()
tangency_portfolio_annualised_stdev = tangency_portfolio_variance**0.5 * 12**0.5
tangency_portfolio_sharpe = tangency_portfolio_annualised_mean/tangency_portfolio_annualised_stdev
print(f"Annualised mean returns for tangency portfolio = {tangency_portfolio_annualised_mean}")
print(f"Annualised volatility for tangency portfolio = {tangency_portfolio_annualised_stdev}")
print(f"Annualised Sharpe for tangency portfolio = {tangency_portfolio_sharpe}")


Annualised mean returns for tangency portfolio = -0.5272759704594937
Annualised volatility for tangency portfolio = 0.3285133339626797
Annualised Sharpe for tangency portfolio = -1.6050367396027652


## Comments

The primary objective of our optimization was to minimize the volatility of the portfolio while aiming for a specified return. We did not implement constraints on short selling, which significantly impacted the results. In the absence of short-selling constraints, the optimization algorithm could assign negative weights to assets with high covariances with others in the portfolio. While the squaring of weights in the volatility formula means that negative weights do not directly reduce volatility, the strategic use of short positions modifies the interaction terms in the portfolio's variance calculation.

In [11]:
# Add constraint that one cann0t short securities, must be

In [12]:
# TIPS are dropped completely from investment set

print("Dropping tips completely:")

data_dropped = data.drop("TIP",axis=1)
stats = summary_stats(data_dropped)
def tangency_portfolio_stats(data_dropped,stats):
    portfolio_mean_excess_returns_df = tangency_portfolio(data=data_dropped)
    merged_df =pd.merge(left=stats,right=portfolio_mean_excess_returns_df,on="Asset", how="outer")
    tangency_portfolio_annualised_mean = merged_df.eval("`Mean returns` * Weights").sum()
    covariance_matrix = data_dropped.cov(ddof=1).to_numpy()
    tangency_portfolio_variance = merged_df["Weights"].to_numpy().T @ covariance_matrix @  merged_df["Weights"].to_numpy()
    tangency_portfolio_annualised_stdev = tangency_portfolio_variance**0.5 * 12**0.5
    tangency_portfolio_sharpe = tangency_portfolio_annualised_mean/tangency_portfolio_annualised_stdev
    print(f"Annualised mean returns for tangency portfolio = {tangency_portfolio_annualised_mean}")
    print(f"Annualised volatility for tangency portfolio = {tangency_portfolio_annualised_stdev}")
    print(f"Annualised Sharpe for tangency portfolio = {tangency_portfolio_sharpe}")
    return {"mean_annualised_return":tangency_portfolio_annualised_mean,
                                    "annualised_volatility":tangency_portfolio_annualised_stdev,
                                    "Sharpe ratio":tangency_portfolio_sharpe}
    
display(tangency_portfolio(data=data_dropped))

tangency_portfolio_stats(data_dropped,stats)

print("No improvement when TIPS are dropped completely")


Dropping tips completely:
Security with best sharpe ratio: SPY


Asset   Weights
0   BWX  3.015911
1   DBC -0.090968
2   EEM -0.664230
3   EFA  0.540606
4   HYG -0.871637
5   IEF -4.832385
6   IYR  0.990156
7   PSP  0.324898
8   QAI  7.976911
9   SPY -5.389260

Annualised mean returns for tangency portfolio = -0.5808055656351949
Annualised volatility for tangency portfolio = 0.3621249970057488
Annualised Sharpe for tangency portfolio = -1.6038814509841046
No improvement when TIPS are dropped completely


In [13]:
# Adjusting tips expected excess return to TIPS is adjusted to be 0.0012 higher than what the historic sample shows

print("Adjusting tips expected excess return to TIPS:")

data_modified = data.copy(deep =True)
data_modified["TIP"]=data.eval("TIP + 0.0012")
stats = summary_stats(data_modified)
display(tangency_portfolio(data=data_modified))

tangency_portfolio_stats(data_modified,stats)


print("No improvement when adjusting expected excess return to TIPS")


Adjusting tips expected excess return to TIPS:
Security with best sharpe ratio: SPY


Asset    Weights
0    BWX  29.260095
1    DBC   1.771624
2    EEM  -5.123028
3    EFA   1.451443
4    HYG  -4.460119
5    IEF -18.604613
6    IYR  11.376147
7    PSP   4.725247
8    QAI  77.462118
9    SPY -49.963644
10   TIP -46.895270

Annualised mean returns for tangency portfolio = -6.174020937258289
Annualised volatility for tangency portfolio = 3.6557676146056397
Annualised Sharpe for tangency portfolio = -1.6888439277681775
No improvement when adjusting expected excess return to TIPS


Does not seem to expand investment set

# 3. Allocations

<i>This section is graded for a good-faith effort by your group. Submit your write-up- along with your supporting code.

* Continue with the same data file as the previous section.

* Suppose the investor has a targeted mean excess return (per month) of $\mutarg$ = 0.01.

Build the following portfolios:

#### Equally-weighted (EW)
Rescale the entire weighting vector to have target mean $\mutarg$. Thus, the $i$ element of the weight vector is,
$$\wEW_i = \frac{1}{n}$$

#### “Risk-parity” (RP)
Risk-parity is a term used in a variety of ways, but here we have in mind setting the weight of the portfolio to be proportional to the inverse of its full-sample variance estimate. Thus, the $i$ element of the weight vector is,
$$\wRP_i = \frac{1}{\sigma_i^2}$$

#### Regularized (REG)
Much like the Mean-Variance portfolio, set the weights proportional to 
$$\wREG \sim \widehat{\Sigma}^{-1}\mux$$
but this time, use a regularized covariance matrix,
$$\widehat{\Sigma} = \frac{\Sigma + \Sigma_D}{2}$$
where $\Sigma_D$ denotes a *diagonal* matrix of the security variances, with zeros in the off-diagonals.

Thus, $\widehat{\Sigma}$ is obtained from the usual covariance matrix, $\Sigma$, but shrinking all the covariances to half their estimated values. 


### Comparing

In order to compare all these allocation methods, (those above, along with the tangency portfolio obtained in the previous section,) rescale each weight vector, such that it has targeted mean return of $\mutarg$.

* Calculate the performance of each of these portfolios over the sample.
* Report their mean, volatility, and Sharpe ratio. 
* How do these compare across the four allocation methods?

In [14]:
target = 0.12
Comparison_dict = {}
weights = tangency_portfolio(data=data)
data_mean = describe(data,axis = 0)[2] * 12
expected_returns = np.reshape(data_mean, (-1,))
data_cov_matrix = data.cov(ddof=1).to_numpy()
inverse_cov_matrix = linalg.inv(data_cov_matrix)
allocation = ((np.ones(shape=(1, inverse_cov_matrix.shape[0])) @ inverse_cov_matrix @ expected_returns)) / ((expected_returns.T @ inverse_cov_matrix @ expected_returns)) * target
weights = (weights.iloc[:,1].to_numpy() * allocation)

portfolio_returns = weights @ expected_returns
annualised_portfolio_variance = weights @ data_cov_matrix @ weights.T * 12
annualised_standard_deviation = annualised_portfolio_variance**0.5
Comparison_dict["MV"]={"mean_annualised_return":portfolio_returns,
                                    "annualised_volatility":annualised_standard_deviation,
                                    "Sharpe ratio":portfolio_returns/annualised_standard_deviation}

print(f"MV portfolio have mean annualised returns of {portfolio_returns}")
print(f"MV portfolio have an annualised volatility of {annualised_standard_deviation}")
print(f"MV portfolio have Sharpe ratio of {portfolio_returns/annualised_standard_deviation}")

MV portfolio have mean annualised returns of 0.11999999999999994
MV portfolio have an annualised volatility of 0.07476464372379354
MV portfolio have Sharpe ratio of 1.6050367396027654


In [15]:
# Equally Weighted portfolio
target = 0.12
securities = len(data.columns)
data_cov_matrix = data.cov().to_numpy()
weights = np.full(securities, 1/securities) 
# portfolio_mean = weights.T @ data[]
expected_returns = data.mean(axis=0).to_numpy()
portfolio_returns = weights @ expected_returns * 12
#scaling
weights = (target/portfolio_returns) * weights
portfolio_returns = weights @ expected_returns * 12
annualised_portfolio_variance = weights @ data_cov_matrix @ weights.T * 12
annualised_standard_deviation = annualised_portfolio_variance**0.5
Comparison_dict["equally_weighted"]={"mean_annualised_return":portfolio_returns,
                                    "annualised_volatility":annualised_standard_deviation,
                                    "Sharpe ratio":portfolio_returns/annualised_standard_deviation}

print(f"Equally weighted portfolio have mean annualised returns of {portfolio_returns}")
print(f"Equally weighted portfolio have an annualised volatility of {annualised_standard_deviation}")
print(f"Equally weighted portfolio have Sharpe ratio of {portfolio_returns/annualised_standard_deviation}")

Equally weighted portfolio have mean annualised returns of 0.11999999999999998
Equally weighted portfolio have an annualised volatility of 0.2915272357323571
Equally weighted portfolio have Sharpe ratio of 0.41162534848088284


In [16]:
# Risk Parity
weights_risk_parity = (1 / (data_cov_matrix.diagonal()*12)) / (1/(data_cov_matrix.diagonal()*12)).sum()

expected_returns = data.mean(axis=0).to_numpy()
portfolio_returns = weights_risk_parity @ expected_returns * 12
weights_risk_parity = (target/portfolio_returns) * weights_risk_parity
portfolio_returns = weights_risk_parity @ expected_returns * 12
annualised_portfolio_variance =weights_risk_parity @ data_cov_matrix @ weights_risk_parity.T * 12
annualised_standard_deviation = annualised_portfolio_variance**0.5
Comparison_dict["risk_parity"]={"mean_annualised_return":portfolio_returns,
                                    "annualised_volatility":annualised_standard_deviation,
                                    "Sharpe ratio":portfolio_returns/annualised_standard_deviation}
print(f"Risk Parity portfolio have mean annualised returns of {portfolio_returns}")
print(f"Risk Parity portfolio have an annualised volatility of {annualised_standard_deviation}")
print(f"Risk Parity portfolio have Sharpe ratio of {portfolio_returns/annualised_standard_deviation}")

Risk Parity portfolio have mean annualised returns of 0.11999999999999998
Risk Parity portfolio have an annualised volatility of 0.3095148749611095
Risk Parity portfolio have Sharpe ratio of 0.3877034989516351


In [17]:
variance_matrix = np.zeros_like(data_cov_matrix)  
np.fill_diagonal(variance_matrix, np.diag(data_cov_matrix)) 
regularized_cov_matrix = (variance_matrix + data_cov_matrix)/2
expected_returns = data.mean(axis=0).to_numpy()
weights = linalg.inv(regularized_cov_matrix) @ expected_returns
portfolio_returns = weights @ expected_returns * 12
#scaling
weights = (target/portfolio_returns) * weights
portfolio_returns = weights @ expected_returns * 12
annualised_portfolio_variance = weights @ data_cov_matrix @ weights.T * 12
annualised_standard_deviation = annualised_portfolio_variance**0.5
Comparison_dict["regularized"]={"mean_annualised_return":portfolio_returns,
                                    "annualised_volatility":annualised_standard_deviation,
                                    "Sharpe ratio":portfolio_returns/annualised_standard_deviation}

print(f"regularized portfolio have mean annualised returns of {portfolio_returns}")
print(f"regularized portfolio have an annualised volatility of {annualised_standard_deviation}")
print(f"regularized portfolio have Sharpe ratio of {portfolio_returns/annualised_standard_deviation}")

regularized portfolio have mean annualised returns of 0.12
regularized portfolio have an annualised volatility of 0.10260435914819349
regularized portfolio have Sharpe ratio of 1.1695409531936323


In [18]:
display(pd.DataFrame(Comparison_dict).transpose())

mean_annualised_return  annualised_volatility  Sharpe ratio
MV                                  0.12               0.074765      1.605037
equally_weighted                    0.12               0.291527      0.411625
risk_parity                         0.12               0.309515      0.387703
regularized                         0.12               0.102604      1.169541

MV Portfolio exhibits the lowest annualized volatility at 0.006230, which translates into the highest Sharpe ratio of 1.605037 among the portfolios. This indicates superior risk-adjusted returns, suggesting that the MV optimization effectively balances return against risk. Equally Weighted Portfolio and Risk Parity Portfolio show higher volatilities at 0.024294 and 0.025793, respectively, with correspondingly lower Sharpe ratios of 0.411625 and 0.387703. These methods distribute risks differently: equally weighted across all assets without considering individual asset volatilities or correlations, and proportionally based on inverse volatility, which does not always result in optimal risk-adjusted outcomes. Regularized Portfolio strikes a balance with a volatility of 0.008550 and a Sharpe ratio of 1.169541. The regularization likely introduces constraints or shrinks estimates to prevent overfitting or excessive trading on noise, thus improving the stability of the risk-return trade-off compared to more naive allocations.

 When comparing these methods, the MV and Regularized portfolios stand out for their higher Sharpe ratios, indicating more efficient use of risk for the targeted return. The MV portfolio, in particular, demonstrates the value of optimization based on the covariance structure and expected returns, achieving the lowest volatility and highest Sharpe ratio.

***

# 4. Out-of-Sample Performance

<i>This section is not graded, and you do not need to submit it. Still, we may discuss it in class, in which case, you would be expected to know it.

## 1. One-step Out-of-Sample (OOS) Performance
Let’s divide the sample to both compute a portfolio and then check its performance out of sample.
* Using only data through the end of `2022`, compute the weights built in Section 3.
* Rescale the weights, (using just the in-sample data,) to set each allocation to have the same mean return of $\mutarg$.
* Using those weights, calculate the portfolio’s Sharpe ratio within that sample.
* Again using those weights, (derived using data through `2022`,) calculate the portfolio’s OOS Sharpe ratio, which is based only on performance in `2023-2024`.

In [19]:
filtered_data = data.loc["2011-02-28":"2022-12-31",:]

annualised_expected_returns = filtered_data.mean(axis = 0).to_numpy()
cov_matrix = filtered_data.cov().to_numpy()
inverse_cov_matrix = linalg.inv(cov_matrix)
    
def portfolio_statistics(data: pd.DataFrame, portfolio: str = "MV", target_return: float = 0.01, weights: np.array = None):
        
    annualised_expected_returns = data.mean(axis = 0).to_numpy() * 12
    cov_matrix = data.cov().to_numpy() * 12 # annualised cov matrix
    inverse_cov_matrix = linalg.inv(cov_matrix)
    
    def scale_weights(weights, target_return = target_return,annualised_expected_returns=annualised_expected_returns):
        portfolio_return = weights.T @ annualised_expected_returns
        scaling_factor = target_return / portfolio_return
        weights = scaling_factor * weights
        portfolio_return = weights.T @ annualised_expected_returns

        if abs(portfolio_return - target_return) <= target_return*0.001:
            return weights
        else:
            raise ValueError("Issues with scaling")
    if weights is not None:
         pass
    elif portfolio == "MV":
        ones = np.ones_like(annualised_expected_returns).T
        tangency_portfolio = inverse_cov_matrix @  annualised_expected_returns / (ones@inverse_cov_matrix@annualised_expected_returns)
        weights = ((ones @ inverse_cov_matrix @ annualised_expected_returns) / (annualised_expected_returns.T @ inverse_cov_matrix @ annualised_expected_returns)) * target_return * tangency_portfolio
    elif portfolio == "EW":
        weights = 1/(len(data.columns)) * np.ones_like(annualised_expected_returns)
        weights = scale_weights(weights=weights)
    
    elif portfolio == "RP":
        individual_var = cov_matrix.diagonal()
        weights = 1/individual_var
        weights = scale_weights(weights=weights)
        
    elif portfolio == "Reg":
        variance_matrix = np.zeros_like(cov_matrix)  
        np.fill_diagonal(variance_matrix, np.diag(cov_matrix)) 
        corrected_cov = (variance_matrix + cov_matrix)/2
        weights = corrected_cov @ annualised_expected_returns
        weights = scale_weights(weights=weights)
    else:
         raise NameError("Unrecognised portfolio optimisation symbol")
    portfolio_return = weights.T @ annualised_expected_returns
    portfolio_variance = weights.T @ cov_matrix @ weights
    portfolio_sharpe = portfolio_return / portfolio_variance**0.5
    return ({"Portfolio": portfolio,
                "Annualised_return":portfolio_return,
                "Annualised volatility": portfolio_variance**0.5,
                "Sharpe": portfolio_sharpe},weights)



MV_weights = portfolio_statistics(data = filtered_data,
                     portfolio="MV",
                     target_return=0.01)
EW_weights = portfolio_statistics(data = filtered_data,
                     portfolio="EW",
                     target_return=0.01)
RP_weights = portfolio_statistics(data = filtered_data,
                     portfolio="RP",
                     target_return=0.01)
Reg_weights = portfolio_statistics(data = filtered_data,
                     portfolio="Reg",
                     target_return=0.01)

In [20]:
# OOS Sharpe ratio

data_OOS = data.loc["2022-12-31":,:]
OOS = [portfolio_statistics(data=data_OOS, portfolio= i[0]["Portfolio"],weights = i[1]) for i in [MV_weights, EW_weights, RP_weights, Reg_weights]]

print("OOS results")
display(pd.DataFrame([i[0] for i in OOS]))


OOS results


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.005498               0.008939  0.615077
1        EW           0.015219               0.028681  0.530640
2        RP           0.010195               0.034518  0.295356
3       Reg           0.018721               0.025370  0.737918

## 2. Rolling OOS Performance

Iterate the Out-of-Sample performance every year, not just the final year. Namely,
* Start at the end of `2015`, and calculate the weights through that time. Rescale them using the mean returns through that time.
* Apply the weights to the returns in the upcoming year, (`2016`.)
* Step forward a year in time, and recompute.
* Continue until again calculating the weights through `2023` and applying them to the returns in `2024`.

Report the mean, volatility, and Sharpe from this dynamic approach for the following portfolios:
* mean-variance (tangency)
* equally-weighted
* risk-parity
* regularized

In [21]:
data.index

DatetimeIndex(['2011-02-28', '2011-03-31', '2011-04-30', '2011-05-31',
               '2011-06-30', '2011-07-31', '2011-08-31', '2011-09-30',
               '2011-10-31', '2011-11-30',
               ...
               '2023-12-31', '2024-01-31', '2024-02-29', '2024-03-31',
               '2024-04-30', '2024-05-31', '2024-06-30', '2024-07-31',
               '2024-08-31', '2024-09-30'],
              dtype='datetime64[ns]', name='Date', length=164, freq=None)

In [22]:
def time_series_split(data, train_start_date,test_length_months=12):
    train_start_date = pd.to_datetime(train_start_date)
    train_end_date = train_start_date + pd.DateOffset(months=test_length_months)
    test_end_date = train_end_date + pd.DateOffset(months=test_length_months)

    filtered_data = data.loc[train_start_date:train_end_date]
    data_OOS = data.loc[train_end_date + pd.DateOffset(days=1):test_end_date]

    return filtered_data, data_OOS

start_date = pd.to_datetime('2015-12-31')
last_date = pd.to_datetime('2024-09-30')
results = []
while start_date + pd.DateOffset(years=1) <= last_date:
    filtered_data, data_OOS = time_series_split(data=data, train_start_date=start_date,
                                                test_length_months=12)

    annualised_expected_returns = filtered_data.mean(axis = 0).to_numpy()
    cov_matrix = filtered_data.cov().to_numpy()
    inverse_cov_matrix = linalg.inv(cov_matrix)

    MV_weights = portfolio_statistics(data = filtered_data,
                        portfolio="MV",
                        target_return=0.01)
    EW_weights = portfolio_statistics(data = filtered_data,
                        portfolio="EW",
                        target_return=0.01)
    RP_weights = portfolio_statistics(data = filtered_data,
                        portfolio="RP",
                        target_return=0.01)
    Reg_weights = portfolio_statistics(data = filtered_data,
                        portfolio="Reg",
                        target_return=0.01)

    OOS = [portfolio_statistics(data=data_OOS, portfolio= i[0]["Portfolio"],weights = i[1]) for i in [MV_weights, EW_weights, RP_weights, Reg_weights]]

    print(f"OOS results from training on {str(start_date.date())} to {str((start_date + pd.DateOffset(years=1)).date())}")
    display(pd.DataFrame([i[0] for i in OOS]))
    start_date = start_date + pd.DateOffset(years=1)

OOS results from training on 2015-12-31 to 2016-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV          -0.004128               0.002548 -1.620196
1        EW           0.023853               0.003923  6.080023
2        RP           0.020395               0.004419  4.614949
3       Reg           0.024146               0.004133  5.842187

OOS results from training on 2016-12-31 to 2017-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV          -0.002810               0.006791 -0.413829
1        EW          -0.006498               0.005772 -1.125698
2        RP          -0.007360               0.006164 -1.194107
3       Reg          -0.006746               0.005974 -1.129159

OOS results from training on 2017-12-31 to 2018-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.003727               0.005933  0.628174
1        EW          -0.022039               0.011449 -1.924924
2        RP          -0.025293               0.010349 -2.444065
3       Reg          -0.020197               0.012439 -1.623650

OOS results from training on 2018-12-31 to 2019-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.008849               0.010235  0.864608
1        EW           0.008874               0.018154  0.488795
2        RP           0.012602               0.012453  1.011909
3       Reg           0.008437               0.021245  0.397129

OOS results from training on 2019-12-31 to 2020-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.013693               0.005107  2.681217
1        EW           0.011120               0.006287  1.768799
2        RP           0.001406               0.004376  0.321310
3       Reg           0.013357               0.007326  1.823123

OOS results from training on 2020-12-31 to 2021-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.003128               0.005579  0.560795
1        EW          -0.011626               0.010915 -1.065083
2        RP          -0.030953               0.024412 -1.267962
3       Reg          -0.006633               0.007514 -0.882792

OOS results from training on 2021-12-31 to 2022-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV          -0.000907               0.005687 -0.159452
1        EW          -0.005694               0.010019 -0.568379
2        RP          -0.002705               0.007356 -0.367671
3       Reg          -0.007097               0.008696 -0.816077

OOS results from training on 2022-12-31 to 2023-12-31


Portfolio  Annualised_return  Annualised volatility    Sharpe
0        MV           0.002921               0.003706  0.787981
1        EW           0.021481               0.017032  1.261229
2        RP           0.034596               0.037045  0.933902
3       Reg           0.013110               0.009259  1.415941

***

# 5. Without a Riskless Asset

<i>This section is not graded, and you do not need to submit it. Still, we may discuss it in class, in which case, you would be expected to know it.

Re-do Section 2 above, but in the model without a risk-free rate.

That is, build the MV allocation using the two-part formula in the `Mean-Variance` section of the notes.
* This essentially substitutes the risk-free rate with the minimum-variance portfolio.
* Now, the allocation depends nonlinearly on the target mean return, $\mutarg$. (With a risk-free rate, we simply scale the weights up and down to achieve the mean return.)

You will find that, conceptually, the answers are very similar. 

In [23]:
data.head()

BWX       DBC       EEM       EFA       HYG       IEF  \
Date                                                                     
2011-02-28  0.006527  0.041620 -0.000527  0.035408  0.014264 -0.002174   
2011-03-31  0.008255  0.026409  0.062723 -0.024055  0.000251 -0.001719   
2011-04-30  0.048261  0.045015  0.026783  0.055715  0.015432  0.017868   
2011-05-31 -0.013447 -0.051625 -0.029301 -0.021962  0.001432  0.025156   
2011-06-30 -0.000270 -0.042817 -0.009528 -0.012280 -0.005878 -0.005217   

                 IYR       PSP       QAI       SPY       TIP  
Date                                                          
2011-02-28  0.045113  0.040056  0.002493  0.034647  0.007118  
2011-03-31 -0.011107  0.016495  0.005348 -0.000052  0.011730  
2011-04-30  0.046089  0.058127  0.018490  0.028418  0.023236  
2011-05-31  0.010233 -0.041465  0.000099 -0.011115  0.002758  
2011-06-30 -0.031233 -0.042754 -0.010948 -0.017042  0.007377

In [24]:
cov_matrix = data.cov().to_numpy()
inverse_cov_matrix = linalg.inv(cov_matrix)
annualised_expected_returns = data.mean(axis=0).to_numpy() * 12
tangency_portfolio = (inverse_cov_matrix @ annualised_expected_returns) / (np.ones_like(annualised_expected_returns).T @ inverse_cov_matrix @ annualised_expected_returns)
mvp = inverse_cov_matrix @ np.ones_like(annualised_expected_returns) / (np.ones_like(annualised_expected_returns).T @ inverse_cov_matrix @ np.ones_like(annualised_expected_returns))

# Initial Allocations
allocation = 0.1
allocation_2 = allocation + 0.001

# Initial Portfolio Returns Calculation
weight = tangency_portfolio * allocation + mvp * (1 - allocation)
weight_2 = tangency_portfolio * allocation_2 + mvp * (1 - allocation_2)
portfolio_return = weight.T @ annualised_expected_returns
portfolio_return_2 = weight_2.T @ annualised_expected_returns

target = 0.01
tolerance = 0.0001
while abs(portfolio_return - target) > tolerance:
    derivative = (portfolio_return_2 - portfolio_return) / (allocation_2 - allocation)
    if derivative == 0:
        break  # Prevent division by zero
    update = (portfolio_return - target) / derivative
    allocation -= update  # Ensure this update is not too large; consider using a damping factor if necessary
    weight = tangency_portfolio * allocation + mvp * (1 - allocation)
    portfolio_return, portfolio_return_2 = portfolio_return_2, weight.T @ annualised_expected_returns
portfolio_return



portfolio_returns = weights @ annualised_expected_returns
annualised_portfolio_variance = weights @ data_cov_matrix @ weights.T * 12
annualised_standard_deviation = annualised_portfolio_variance**0.5
Comparison_dict["MV"]={"mean_annualised_return":portfolio_returns,
                                    "annualised_volatility":annualised_standard_deviation,
                                    "Sharpe ratio":portfolio_returns/annualised_standard_deviation}

print(f"MV portfolio have mean annualised returns of {portfolio_returns}")
print(f"MV portfolio have an annualised volatility of {annualised_standard_deviation}")
print(f"MV portfolio have Sharpe ratio of {portfolio_returns/annualised_standard_deviation}")


MV portfolio have mean annualised returns of 0.12
MV portfolio have an annualised volatility of 0.10260435914819349
MV portfolio have Sharpe ratio of 1.1695409531936323


***